02_knight_and_princess/novikov_dmitry/graph.hpp

```diff
@@ -0,0 +1,325 @@
+                       const Edge::Color& color);
+
+class Graph {
+ public:
+  VertexId add_vertex() {
+    const auto new_vertex_id = get_default_vertex_id();
+    vertex_map_.insert({new_vertex_id, Vertex(new_vertex_id)});
+    depth_map_.at(DEFAULT_DEPTH).push_back(new_vertex_id);
+    return new_vertex_id;
+  }
+
+  void set_vertex_depth(const VertexId& from_vertex_id,
+                        const VertexId& to_vertex_id) {
+    assert(has_vertex(from_vertex_id) && "Vertex doesn't exists");
+    assert(has_vertex(to_vertex_id) && "Vertex doesn't exists");
+    assert(check_binding(from_vertex_id, to_vertex_id) &&
+           "Vertices already binded");
+    VertexId parent_vertex_id = std::max(from_vertex_id, to_vertex_id);
+    VertexId son_vertex_id = std::min(from_vertex_id, to_vertex_id);
+
+    const auto new_son_vertex_depth =
+        vertex_map_.at(parent_vertex_id).depth + 1;
+
+    auto& son_vertex = vertex_map_.at(son_vertex_id);
+    auto& depth_map_son_level = depth_map_.at(son_vertex.depth);
+
+    const auto vertex_iter_at_depth_map = std::remove(
+        depth_map_son_level.begin(), depth_map_son_level.end(), son_vertex.id);
+    depth_map_son_level.erase(vertex_iter_at_depth_map);
+    son_vertex.depth = new_son_vertex_depth;
+
+    if (depth_map_.size() <= new_son_vertex_depth) {
+      depth_map_.push_back(std::vector<VertexId>({son_vertex.id}));
+    } else {
+      depth_map_.at(new_son_vertex_depth).push_back(son_vertex.id);
+    }
+  }
+
+  void add_edge(const VertexId& from_vertex_id,
+                const VertexId& to_vertex_id,
+                const Edge::Color& new_edge_color = Edge::Color::Gray) {
+    assert(has_vertex(from_vertex_id) && "Vertex doesn't exists");
+    assert(has_vertex(to_vertex_id) && "Vertex doesn't exists");
+    assert(!check_binding(from_vertex_id, to_vertex_id) &&
+           "Vertices already binded");
+    assert(check_color_valid(this->get_vertex(from_vertex_id),
+                             this->get_vertex(to_vertex_id), new_edge_color) &&
+           "Not valid color");
+    const auto new_edge_id = get_default_edge_id();
+    const auto new_edge =
+        edge_map_.insert({new_edge_id, Edge(from_vertex_id, to_vertex_id,
+                                            new_edge_id, new_edge_color)});
+    vertex_map_.at(from_vertex_id).add_edge_id(new_edge.first->first);
+    if (from_vertex_id != to_vertex_id) {
+      vertex_map_.at(to_vertex_id).add_edge_id(new_edge.first->first);
+    }
+    if (new_edge_color == Edge::Color::Gray) {
+      set_vertex_depth(from_vertex_id, to_vertex_id);
+    }
+  }
+
+  bool check_binding(const VertexId& from_vertex_id,
+                     const VertexId& to_vertex_id) const {
+    assert(has_vertex(from_vertex_id) && "Vertex doesn't exists");
+    assert(has_vertex(to_vertex_id) && "Vertex doesn't exists");
+    const auto& from_vertex = this->get_vertex(from_vertex_id);
+    const auto& to_vertex = this->get_vertex(to_vertex_id);

```

Comment:
зачем `this->`? Без этого разве не работает?
```suggestion
    const auto& from_vertex = get_vertex(from_vertex_id);
    const auto& to_vertex = get_vertex(to_vertex_id);
```